src/ccpp_constituent_prop_mod.F90

```diff
@@ -914,351 +1283,975 @@ subroutine ccp_model_const_lock(this, ncols, num_layers, num_interfaces,   &
+               call this%const_metadata(index)%deallocate()
+            end do
+         end if
+         deallocate(this%const_metadata)
+      end if
+      if (clear_table) then
+         this%num_layer_vars = 0
+         this%num_advected_vars = 0
+         this%num_layers = 0
+         call this%hash_table%clear()
+      end if
+
+   end subroutine ccp_model_const_reset
+
+   !########################################################################
+
+   logical function ccp_model_const_is_match(this, index, advected,            &
+        thermo_active) result(is_match)
+      ! Return .true. iff the constituent at <index> matches a pattern
+      ! Each (optional) property which is present represents something
+      !    which is required as part of a match.
+      ! Since this is a private function, error checking for locked status
+      !    is *not* performed.
+
+      ! Dummy arguments
+      class(ccpp_model_constituents_t), intent(in) :: this
+      integer,                          intent(in) :: index
+      logical, optional,                intent(in) :: advected
+      logical, optional,                intent(in) :: thermo_active
+      ! Local variable
+      logical :: check
+
+      ! By default, every constituent is a match
+      is_match = .true.
+      if (present(advected)) then
+         call this%const_metadata(index)%is_advected(check)
+         if (advected .neqv. check) then
+            is_match = .false.
+         end if
+      end if
+
+      if (present(thermo_active)) then
+         call this%const_metadata(index)%is_thermo_active(check)
+         if (thermo_active .neqv. check) then
+            is_match = .false.
+         end if
+      end if
+
+
+   end function ccp_model_const_is_match
+
+   !########################################################################
+
+   subroutine ccp_model_const_num_match(this, nmatch, advected, thermo_active, &
+        errcode, errmsg)
+      ! Query number of constituents matching pattern
+      ! Each (optional) property which is present represents something
+      !    which is required as part of a match.
+      ! <this> must be locked to execute this function
+
+      ! Dummy arguments
+      class(ccpp_model_constituents_t), intent(in)  :: this
+      integer,                          intent(out) :: nmatch
+      logical,          optional,       intent(in)  :: advected
+      logical,          optional,       intent(in)  :: thermo_active
+      integer,          optional,       intent(out) :: errcode
+      character(len=*), optional,       intent(out) :: errmsg
+      ! Local variables
+      integer                     :: index
+      character(len=*), parameter :: subname = "ccp_model_const_num_match"
+
+      nmatch = 0
+      if (this%const_props_locked(errcode=errcode, errmsg=errmsg, warn_func=subname)) then
+         do index = 1, SIZE(this%const_metadata)
+            if (this%is_match(index, advected=advected, thermo_active=thermo_active)) then
```

Comment:
No need, the Fortran standard guarantees proper behavior given that these items are optional inputs to `is_match`.

Reply:
BTW, this extends to `allocatable` arrays and `pointers`. If you pass an unallocated array to an ordinary, `optional` dummy argument, it will be `.not. present`. Same if you pass an `unassociated` pointer.
Example:
```
deallocate(foo)
call bar(opt_arg=foo)
```
Inside `bar`, `opt_arg` will be `.not. present`.